Refactor planner code's handling of sort ordering in Motion nodes

hlinnaka opened the pull request:
@hsyuan asked me to have a quick look at PR #1450. I'm not entirely sure why it didn't work as intended, but it got me thinking, why we need to track sort ordering in Flow nodes in the first place. It seems error-prone, having to keep the column index information up-to-date throughout the later stages of the planner, given that PostgreSQL doesn't need to do that. Digging into that rabbit hole, I ended up with this patch.

bhuvnesh2703 commented:
Testing with this fix, we observed the below failures in table_function.

1. Redistribute motion is removed from the plan after this fix and it produces wrong results
On Master:
```
pivotal=# explain SELECT * FROM table_function.multiset_5( TABLE (SELECT * FROM table_function.example ORDER BY b, a SCATTER RANDOMLY) );
                                             QUERY PLAN
-----------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice2; segments: 3)  (cost=2.27..2.42 rows=10 width=36)
   ->  Table Function Scan on multiset_5  (cost=2.27..2.42 rows=4 width=36)
         ->  Sort  (cost=2.27..2.29 rows=4 width=16)
               Sort Key: example.b, example.a
               ->  Redistribute Motion 3:3  (slice1; segments: 3)  (cost=0.00..2.10 rows=4 width=16)
                     ->  Seq Scan on example  (cost=0.00..2.10 rows=4 width=16)
 Optimizer status: legacy query optimizer
(7 rows)

pivotal=# SELECT * FROM table_function.multiset_5( TABLE (SELECT * FROM table_function.example ORDER BY b, a SCATTER RANDOMLY) );
 a |      b
---+-------------
 1 |  value1.2/4
 2 |  value2.3/3
 1 |  value1.1/4
 1 |  value1.4/4
 3 |  value3.1/2
 4 |  value4.1/1
 1 |  value1.3/4
 2 |  value2.1/3
 2 |  value2.2/3
 3 |  value3.2/2
(10 rows)
```

With this fix:
```
gpadmin=# SELECT * FROM table_function.multiset_5( TABLE (SELECT * FROM table_function.example ORDER BY b, a SCATTER RANDOMLY) );
 a |     b
---+------------
 3 | value3.1/2
 3 | value3.2/2
 4 | value4.1/1
 1 | value1.1/4
 1 | value1.2/4
 1 | value1.3/4
 1 | value1.4/4
 2 | value2.1/3
 2 | value2.2/3
 2 | value2.3/3
(10 rows)

gpadmin=# explain SELECT * FROM table_function.multiset_5( TABLE (SELECT * FROM table_function.example ORDER BY b, a SCATTER RANDOMLY) );
                                  QUERY PLAN
------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=2.63..2.93 rows=20 width=36)
   ->  Table Function Scan on multiset_5  (cost=2.63..2.93 rows=7 width=36)
         ->  Sort  (cost=2.63..2.68 rows=7 width=16)
               Sort Key: example.b, example.a
               ->  Seq Scan on example  (cost=0.00..2.20 rows=7 width=16)
 Optimizer status: legacy query optimizer
(6 rows)
```

2. Looks like this fix attaches the tupdesc but the results are wrong.

On Master
```
  select q from (select max(f1) from int4_tbl group by f1 order by f1) q;
ERROR:  record type has not been registered
```

With this fix
```
  select q from (select max(f1) from int4_tbl group by f1 order by f1) q;
        q
 ----------------
  (-2147483647,)
  (-123456,)
  (0,)
  (123456,)
  (2147483647,)
 (5 rows)
```

3.  Table Function scan is pulled up before Gather Motion
On Master:
```
pivotal=# explain SELECT * FROM multiset_5( TABLE (SELECT * FROM example ORDER BY a, b) );
                                  QUERY PLAN
------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=2.27..2.42 rows=10 width=36)
   ->  Table Function Scan on multiset_5  (cost=2.27..2.42 rows=4 width=36)
         ->  Sort  (cost=2.27..2.29 rows=4 width=16)
               Sort Key: example.a, example.b
               ->  Seq Scan on example  (cost=0.00..2.10 rows=4 width=16)
 Optimizer status: legacy query optimizer
```
With this fix
```
gpadmin=# explain SELECT * FROM multiset_5( TABLE (SELECT * FROM example ORDER BY a, b) );
                                     QUERY PLAN
------------------------------------------------------------------------------------
 Table Function Scan on multiset_5  (cost=2.63..2.93 rows=20 width=36)
   ->  Gather Motion 3:1  (slice1; segments: 3)  (cost=2.63..2.68 rows=20 width=16)
         Merge Key: example.a, example.b
         ->  Sort  (cost=2.63..2.68 rows=7 width=16)
               Sort Key: example.a, example.b
               ->  Seq Scan on example  (cost=0.00..2.20 rows=7 width=16)
 Optimizer status: legacy query optimizer
```

oarap commented:
I an @hsyuan think this PR only requires to update the test files and it is good to go.

vraghavan78 commented:
One of them looks likes the answer changes. Is that okay?

bhuvnesh2703 commented:
@oarap @hsyuan the results for the 2 issue above are incorrect as they are has an extra `,`. From postgresql results:

```
pivotal=# create table int4_tbl (f1 int);
CREATE TABLE
pivotal=# insert into int4_tbl values(123456), (-2147483647), (0), (-123456), (2147483647);
INSERT 0 5
pivotal=# select q from (select max(f1) from int4_tbl group by f1 order by f1) q;
       q
---------------
 (-2147483647)
 (-123456)
 (0)
 (123456)
 (2147483647)
(5 rows)
```

Eulerizeit commented:
What is causing that?

-- Rob

On Fri, Jan 20, 2017 at 9:06 PM, Bhuvnesh <notifications@github.com> wrote:

> @oarap <https://github.com/oarap> @hsyuan <https://github.com/hsyuan> the
> results for the 2 issue above are incorrect as they are has an extra ,.
> From postgresql results:
>
> pivotal=# create table int4_tbl (f1 int);
> CREATE TABLE
> pivotal=# insert into int4_tbl values(123456), (-2147483647 <(214)%20748-3647>), (0), (-123456), (2147483647);
> INSERT 0 5
> pivotal=# select q from (select max(f1) from int4_tbl group by f1 order by f1) q;
>        q
> ---------------
>  (-2147483647)
>  (-123456)
>  (0)
>  (123456)
>  (2147483647)
> (5 rows)
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/greenplum-db/gpdb/pull/1460#issuecomment-274224972>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AJOIZ-dNOfyVx2YyZKX1epaXuJCgVn9Iks5rUWhDgaJpZM4LREI7>
> .
>

dhanashreek89 commented:
https://github.com/greenplum-db/gpdb/issues/618 addresses issue #2 described above.  We will remove that test case for now; it can be added when we have a fix for this issue.

The other tests will be updated to pass the ICG.

Eulerizeit commented:
Dhanashree,

Can you clarify what you mean here. I don't understand why you are removing
a test.



On Mon, Jan 23, 2017 at 7:07 PM, Dhanashree Kashid <notifications@github.com
> wrote:

> #618 <https://github.com/greenplum-db/gpdb/issues/618> addresses issue #2
> <https://github.com/greenplum-db/gpdb/pull/2> described above. We will
> remove that test case for now; it can be added when we have a fix for this
> issue.
>
> The other tests will be updated to pass the ICG.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/greenplum-db/gpdb/pull/1460#issuecomment-274659205>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AJOIZ6uD3JU9XrkkmWOn-KmqFCzomuLwks5rVUDEgaJpZM4LREI7>
> .
>

dhanashreek89 commented:
Robert, 

This test case is been tracked as an issue and it seems like it needs to be addressed differently other than the fix introduced with this pull request. If we merge this fix with this test case, the result is already wrong and fixing it is beyond the scope of this pull request. That's why we are proposing to remove it temporarily until the issue #618 is resolved. 

What is your input on this @hlinnaka ?

Dhanashree and Omer

Eulerizeit commented:
I guess I don't understand. This seems like a bit of a rabbit hole. I'm ok
with holding off on fixing something. What I'm hesitant about is removing a
test. Also, since we are removing a test, how is the CI green right now?



On Mon, Jan 23, 2017 at 7:42 PM, Dhanashree Kashid <notifications@github.com
> wrote:

> Robert,
>
> This test case is been tracked as an issue and it seems like it needs to
> be addressed differently other than the fix introduced with this pull
> request. If we merge this fix with this test case, the result is already
> wrong and fixing it is beyond the scope of this pull request. That's why we
> are proposing to remove it temporarily until the issue #618
> <https://github.com/greenplum-db/gpdb/issues/618> is resolved.
>
> What is your input on this @hlinnaka <https://github.com/hlinnaka> ?
>
> Dhanashree and Omer
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/greenplum-db/gpdb/pull/1460#issuecomment-274665641>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AJOIZxOCcuPztnqZTR3SDTNctSmm37v2ks5rVUkOgaJpZM4LREI7>
> .
>

dhanashreek89 commented:
Currently, the expected result file for this query, has ERROR :

expected/subselect.out
```
select q from (select max(f1) from int4_tbl group by f1 order by f1) q;
ERROR:  record type has not been registered
```

Hence CI is green.

ashwinstar commented:
@vraghavan78 what's the plan for this PR ?

vraghavan78 commented:
@ashwinstar this is a dev issue. @hlinnaka / @dhanashreek89 can you reply to @ashwinstar concern.

hlinnaka commented:
I rebased this over current master.

It looks like the bug that this fixes, and adds a test case for, was already fixed as a side-effect of the planpartition.c rewrite in commit e378d84bf3. I'm not sure if there's some instance of the same bug that could be triggered with a different query.

hlinnaka commented:
Added a commit to fix (I hope) the table_function failure that @bhuvnesh2703 pointed out.

hlinnaka commented:
There are still some failures in the table_functions test. I'll dig into them later, but I think they can be fixed with some fairly small changes, by moving things around a little bit in grouping_planner..

Eulerizeit commented:
@hlinnaka @vraghavan78 what is the state of this PR? Should it be closed? Is the issue it is addressing being dealt with elsewhere?

hsyuan commented:
It should not be closed. I think @hlinnaka is still working on this PR.

hlinnaka commented:
Rebased and fixed a regression test failure.

Note: This changes the Query and Flow structs, which are included in the serialized form of views. That counts as catalog changes, so we cannot push this until the tree is open for catalog changes again.

hsyuan commented:
@hlinnaka Can this PR go to master branch now?

hlinnaka commented:
So, there was one regression failure left in this. It's because of a plan change in following query:

Test schema:
```
create table testsisc (i1 int, i2 int, i3 int, i4 int);
insert into testsisc select i, i % 1000, i % 100000, i % 75 from
	(select generate_series(1, nsegments * 50000) as i from
	(select count(*) as nsegments from gp_segment_configuration where role='p' and content >= 0) foo) bar;
```

GPDB 5:
```
postgres=# explain 
  with ctesisc as (select * from testsisc order by i2)
  select t1.i3, t2.i2
  from ctesisc as t1, ctesisc as t2
  where t1.i1 = t2.i2;
                                             QUERY PLAN                                              
-----------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice2; segments: 3)  (cost=4885.99..293794.28 rows=22601515 width=8)
   ->  Hash Join  (cost=4885.99..293794.28 rows=7533839 width=8)
         Hash Cond: t2.i2 = t1.i1
         ->  Redistribute Motion 3:3  (slice1; segments: 3)  (cost=0.00..6013.52 rows=50113 width=4)
               Hash Key: t2.i2
               ->  Subquery Scan t2  (cost=0.00..3006.76 rows=50113 width=4)
                     ->  Sort  (cost=14634.83..15010.68 rows=50113 width=16)
                           Sort Key: public.testsisc.i2
                           ->  Seq Scan on testsisc  (cost=0.00..1707.38 rows=50113 width=16)
         ->  Hash  (cost=3006.76..3006.76 rows=50113 width=8)
               ->  Subquery Scan t1  (cost=0.00..3006.76 rows=50113 width=8)
                     ->  Sort  (cost=14634.83..15010.68 rows=50113 width=16)
                           Sort Key: public.testsisc.i2
                           ->  Seq Scan on testsisc  (cost=0.00..1707.38 rows=50113 width=16)
 Optimizer status: legacy query optimizer
(15 rows)
```

With this PR:
```
postgres=# explain 
  with ctesisc as (select * from testsisc order by i2)
  select t1.i3, t2.i2
  from ctesisc as t1, ctesisc as t2
  where t1.i1 = t2.i2;
                                                 QUERY PLAN                                                 
------------------------------------------------------------------------------------------------------------
 Hash Join  (cost=4827.58..283975.11 rows=22064429 width=8)
   Hash Cond: t1.i1 = t2.i2
   ->  Subquery Scan t1  (cost=0.00..2970.82 rows=148541 width=8)
         ->  Gather Motion 3:1  (slice1; segments: 3)  (cost=14449.45..14820.81 rows=148541 width=16)
               Merge Key: public.testsisc.i2
               ->  Sort  (cost=14449.45..14820.81 rows=49514 width=16)
                     Sort Key: public.testsisc.i2
                     ->  Seq Scan on testsisc  (cost=0.00..1689.41 rows=49514 width=16)
   ->  Hash  (cost=2970.82..2970.82 rows=49514 width=4)
         ->  Subquery Scan t2  (cost=0.00..2970.82 rows=148541 width=4)
               ->  Gather Motion 3:1  (slice2; segments: 3)  (cost=14449.45..14820.81 rows=148541 width=16)
                     Merge Key: public.testsisc.i2
                     ->  Sort  (cost=14449.45..14820.81 rows=49514 width=16)
                           Sort Key: public.testsisc.i2
                           ->  Seq Scan on testsisc  (cost=0.00..1689.41 rows=49514 width=16)
 Settings:  optimizer=off
 Optimizer status: legacy query optimizer
(17 rows)
```

So, what we have here is:

* A subquery with an ORDER BY
* In GPDB 5, the planner honors the ORDER BY, but sorts separately on each node.
* After this PR, the planner will bring all the data to a single node, because of the ORDER BY.

Strictly speaking, the planner could just discard the ORDER BY in the subquery. Having an ORDER BY in a subquery like that doesn't give any guarantees about the order of the overall query. Indeed, ORCA discards it. But if we don't want to discard it altogether, I think the new behavior makes sense: If a user specified ORDER BY, presumably the rows should be sorted at some point, and "sorted separately on each node" isn't really the thing. Or maybe we should try to preserve the old behavior. However, this is from workfile/sisc_sort_spill test, so if we change the behavior, we need to find a new way to write that spill test case, so that it still tests the same thing.

Now, consider this (another test from our regression suite):

```
postgres=# SELECT ARRAY(select f2 from arrtest_f order by f2) AS "ARRAY" ORDER BY 1; -- MPP-11853
                     ARRAY                     
-----------------------------------------------
 {1.15,1.15,1.18,1.21,1.24,1.26,1.26,1.3,1.32}
(1 row)
```

Does that need to honor the ORDER BY? It would be surprising if it didn't. We have no precedence from ORCA here, because it falls back to the Postgres planner on that query. In fact, looking at MPP-11853, at some point we didn't preserve the order in that query, and we changed it so that we do, and the query was added to the regression suite specifically to test that.

I added some code to preserve the old behavior, and the tests pass now. ARRAY-sublinks are now special-cased, so that we fully honor the ORDER BY in such subqueries, while for all other subqueries, you get a partial, per-node, order. I'm not sure that's the most sensible behavior, but this will do for now, I think.

Let's see what the PR pipeline thinks of this now.

hlinnaka commented:
Ok, all the tests pass now. I think this has sat in the PR queue for long enough, so pushed.
